resources/user.go

```diff
@@ -0,0 +1,306 @@
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *UserRes) Validate() error {
+	if obj.State != "exists" && obj.State != "absent" {
+		return fmt.Errorf("State must be 'exists' or 'absent'")
+	}
+	return obj.BaseRes.Validate()
+}
+
+// Init initializes the resource.
+func (obj *UserRes) Init() error {
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *UserRes) Watch() error {
+	var err error
+	obj.recWatcher, err = recwatch.NewRecWatcher(passwdFile, false)
+	if err != nil {
+		return err
+	}
+	defer obj.recWatcher.Close()
+
+	// notify engine that we're running
+	if err := obj.Running(); err != nil {
+		return err // bubble up a NACK...
+	}
+
+	var send = false // send event?
+	var exit *error
+
+	for {
+		if obj.debug {
+			log.Printf("Watching: %s", passwdFile) // attempting to watch...
+		}
+
+		select {
+		case event, ok := <-obj.recWatcher.Events():
+			if !ok { // channel shutdown
+				return nil
+			}
+			if err := event.Error; err != nil {
+				return errwrap.Wrapf(err, "Unknown %s watcher error", obj)
+			}
+			if obj.debug { // don't access event.Body if event.Error isn't nil
+				log.Printf("%s: Event(%s): %v", obj, event.Body.Name, event.Body.Op)
+			}
+			send = true
+			obj.StateOK(false) // dirty
+
+		case event := <-obj.Events():
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit // exit
+			}
+			//obj.StateOK(false) // dirty // these events don't invalidate state
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			send = false
+			obj.Event()
+		}
+	}
+}
+
+// CheckApply method for User resource.
+func (obj *UserRes) CheckApply(apply bool) (checkOK bool, err error) {
+	log.Printf("%s: CheckApply(%t)", obj, apply)
+
+	var exists = true
+	usr, err := user.Lookup(obj.GetName())
+	if err != nil {
+		if _, ok := err.(user.UnknownUserError); !ok {
+			return false, errwrap.Wrapf(err, "error looking up user")
+		}
+		log.Printf("the user: %s does not exist", obj.GetName())
+		exists = false
+	}
+
+	if obj.AllowDuplicateUID == false && obj.UID != nil {
+		existingUID, err := user.LookupId(strconv.Itoa(int(*obj.UID)))
+		if err != nil {
+			if _, ok := err.(user.UnknownUserIdError); !ok {
+				return false, errwrap.Wrapf(err, "error looking up UID")
+			}
+		} else if existingUID.Username != obj.GetName() {
+			return false, errwrap.Wrapf(err, "the requested UID is already taken")
+		}
+	}
+
+	if usercheck := true; exists && obj.State == "exists" {
+		if obj.GID != nil && strconv.Itoa(int(*obj.GID)) != usr.Gid {
+			usercheck = false
+		}
+		if obj.UID != nil && strconv.Itoa(int(*obj.UID)) != usr.Uid {
+			usercheck = false
+		}
+		if obj.HomeDir != nil && *obj.HomeDir != usr.HomeDir {
+			usercheck = false
+		}
+		if usercheck {
+			return true, nil
+		}
+	}
+
+	if obj.State == "absent" && !exists {
```

Comment:
This is good! However, it's probably better to move this check above the "usercheck" section, because it's cheaper and we can handle the simple case first.